File: golang/binary-tree-general/construct_binary_tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	inorderMap := make(map[int]int)

	for i, val := range inorder {
		inorderMap[val] = i
	}

	return build(preorder, 0, len(preorder)-1, 0, len(inorder)-1, inorderMap)
}

func build(preorder []int, preStart, preEnd, inStart, inEnd int, inorderMap map[int]int) *TreeNode {
	if preStart > preEnd || inStart > inEnd {
		return nil
	}

	rootVal := preorder[preStart]
	root := &TreeNode{Val: rootVal}
	rootIndex := inorderMap[rootVal]
	leftTreeSize := rootIndex - inStart

	root.Left = build(preorder, preStart+1, preStart+leftTreeSize, inStart, rootIndex-1, inorderMap)
	root.Right = build(preorder, preStart+leftTreeSize+1, preEnd, rootIndex+1, inEnd, inorderMap)

	return root
}

File: java/src/main/java/com/sanedge/interview/binary_tree_general/ConstructBinaryTree.java
package com.sanedge.interview.binary_tree_general;

import com.sanedge.interview.utils.TreeNode;
import java.util.HashMap;
import java.util.Map;

public class ConstructBinaryTree {

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> inorderMap = new HashMap<>();

        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }

        return build(
            preorder,
            0,
            preorder.length - 1,
            0,
            inorder.length - 1,
            inorderMap
        );
    }

    private TreeNode build(
        int[] preorder,
        int preStart,
        int preEnd,
        int inStart,
        int inEnd,
        Map<Integer, Integer> inorderMap
    ) {
        if (preStart > preEnd || inStart > inEnd) {
            return null;
        }

        int rootVal = preorder[preStart];
        TreeNode root = new TreeNode(rootVal);

        int rootIndex = inorderMap.get(rootVal);

        int leftTreeSize = rootIndex - inStart;

        root.left = build(
            preorder,
            preStart + 1,
            preStart + leftTreeSize,
            inStart,
            rootIndex - 1,
            inorderMap
        );

        root.right = build(
            preorder,
            preStart + leftTreeSize + 1,
            preEnd,
            rootIndex + 1,
            inEnd,
            inorderMap
        );

        return root;
    }
}
